### ISSUE TITLE
Quota Template with dimensions from request header

### ISSUE THREAD

**Issue by DaspawnW** [istio/istio#8526]
**Describe the bug**
I tried to deploy a redis request quota based on your example https://preliminary.istio.io/docs/tasks/policy-enforcement/rate-limiting.html.
The idea is here to use an authorization header to define as dimension for redisquota.

This doesn't seem to work. It seems that the authorization header is ignored.

**Expected behavior**
If I access the URL in two different requests with two different authorization headers I'll have access if for my authorization header the quota is not exceeded. For the other account with a maxAmount of 1 request in 20s the request should be denied.

**Steps to reproduce the bug**
Here is my configuration
```yaml
apiVersion: "config.istio.io/v1alpha2"
kind: redisquota
metadata:
  name: handler
  namespace: istio-system
spec:
  connectionPoolSize: 10
  redisServerUrl: redis-master.redis:6379
  quotas:
  - name: requestcount.quota.istio-system
    maxAmount: 1
    validDuration: 20s
    bucketDuration: 500ms
    rateLimitAlgorithm: ROLLING_WINDOW
---
apiVersion: "config.istio.io/v1alpha2"
kind: quota
metadata:
  name: requestcount
  namespace: istio-system
spec:
  dimensions:
    source: request.headers["X-Test-Header"] | "unknown"
    destination: destination.labels["app"] | destination.workload.name | "unknown"
---
apiVersion: config.istio.io/v1alpha2
kind: rule
metadata:
  name: quota
  namespace: istio-system
spec:
  # quota only applies if you are not logged in.
  # match: match(request.headers["cookie"], "user=*") == false
  actions:
  - handler: handler.redisquota
    instances:
    - requestcount.quota
---
apiVersion: config.istio.io/v1alpha2
kind: QuotaSpec
metadata:
  name: request-count
  namespace: istio-system
spec:
  rules:
  - quotas:
    - charge: 1
      quota: requestcount
---
apiVersion: config.istio.io/v1alpha2
kind: QuotaSpecBinding
metadata:
  name: request-count
  namespace: istio-system
spec:
  quotaSpecs:
  - name: request-count
    namespace: istio-system
  services:
  - name: ms-1
    namespace: microservices
    #  - service: '*'  # Uncomment this to bind *all* services to request-count
---
```

```bash
root@busybox:/# echo "$(date)" && curl https://ms-1.example.play/api/ms1  --insecure -H "X-Test-Header: 789" && echo ""
Thu Sep  6 14:27:35 UTC 2018
Hello World from ms1
root@busybox:/# echo "$(date)" && curl https://ms-1.example.play/api/ms1  --insecure -H "X-Test-Header: 789" && echo ""
Thu Sep  6 14:27:36 UTC 2018
Hello World from ms1
root@busybox:/# echo "$(date)" && curl https://ms-1.example.play/api/ms1  --insecure -H "X-Test-Header: 789" && echo ""
Thu Sep  6 14:27:37 UTC 2018
RESOURCE_EXHAUSTED:Quota is exhausted for: requestcount
root@busybox:/# echo "$(date)" && curl https://ms-1.example.play/api/ms1  --insecure -H "X-Test-Header: 123" && echo ""
Thu Sep  6 14:27:41 UTC 2018
RESOURCE_EXHAUSTED:Quota is exhausted for: requestcount

```
Here are the calls to access it via an external busybox container:

**Version**
```bash
istioctl version
Version: 1.0.1
GitRevision: 42773aacced474d97159902d20579a25b1f98106
User: root@832d5020b1d4
Hub: gcr.io/istio-release
GolangVersion: go1.10.1
BuildStatus: Clean

kubectl version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.0", GitCommit:"925c127ec6b946659ad0fd596fa959be43f0cc05", GitTreeState:"clean", BuildDate:"2017-12-16T03:16:50Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:08:34Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}

```
**Is Istio Auth enabled or not?**
```
~/Downloads/darwin-amd64/helm template install/kubernetes/helm/istio \
    --name istio \
    --namespace istio-system \
    --set global.mtls.enabled=true \
    --set global.proxy.includeIPRanges="10.32.0.0/24" \ # also include pod ip range to prevent direct access to port via port scan
    --set sidecarInjectorWebhook.enabled=false \
    --set ingress.enabled=false \
    --set gateways.istio-ingressgateway.enabled=true \
    --set gateways.istio-ingressgateway.type=NodePort \
    --set gateways.istio-egressgateway.enabled=true \
    --set gateways.istio-egressgateway.type=NodePort \
    --set tracing.enabled=true \
    --set kiali.enabled=true \
    --set security.enabled=true \
    > ../istio.yaml
```
**Environment**
Ubuntu 16.04

**Comment by mandarjog** [comment]
The rule in your configuration is applying quota unconditionally.
So all requests are going to be subject to the same quota `requestcount.quota.istio-system`, with source and destination dimensions.

If you wish you use the authorization header, you should 
1. dimension quota by `request.headers["authorization"]` and you can also reinstate `match` condition in the rule which is commented out in the config.

**Comment by DaspawnW** [comment]
Hi @mandarjog,

sorry I'm a bit confused.

How do I specify this, does I have to change this
```
  quotas:
  - name: requestcount.quota.istio-system
    maxAmount: 1
    validDuration: 20s
    bucketDuration: 500ms
    rateLimitAlgorithm: ROLLING_WINDOW
```
to
```
  quotas:
  - name:request.headers["authorization"]
    maxAmount: 1
    validDuration: 20s
    bucketDuration: 500ms
    rateLimitAlgorithm: ROLLING_WINDOW
```
?

Why do I need the match condition in the quota, I thought the match in it is only if you will bypass the quota for special users?

**Comment by DaspawnW** [comment]
So to be more general, we are using in istio a JWT-Policy that connects to an keycloak server to validate JWTokens that are send as Authorization header.
Now I would be able to set Quotas per User. First I tried to use the JWToken as unique identifier, I know this was a stupid idea. Now I see that the jwt generates some values like: **request.auth.principal**.
I will do the quota dynamically per principal  because I don't know the users in the Keycloak.

As described in the tutorial https://preliminary.istio.io/docs/tasks/policy-enforcement/rate-limiting/ it can be per Client IP (In the above example we have effectively rate limited productpage at 2 rps per client IP). I copied the example above and adjusted it now to have as source of the dimensions in the quota-CRD the principal:
```
apiVersion: "config.istio.io/v1alpha2"
kind: quota
metadata:
  name: requestcount
  namespace: istio-system
spec:
  dimensions:
    source: request.auth.principal
    destination: destination.labels["app"] | destination.workload.name | "unknown"
```

So I don't understand why this and also my old example with authorization header won't work, when you describe in the tutorial, that it's possible by client ip?

Here is also my auth policy:
```
apiVersion: "authentication.istio.io/v1alpha1"
kind: "Policy"
metadata:
  name: "ms-1"
  namespace: microservices
spec:
  targets:
  - name: ms-1
  peers:
  - mtls:
      mode: STRICT
  origins:
  - jwt:
      issuer: "https://<url>/auth/realms/default"
      jwksUri: "https://<url>/auth/realms/default/protocol/openid-connect/certs"
      audiences:
        - test-api
  principalBinding: USE_ORIGIN
```

**Comment by mandarjog** [comment]
Is no quota being applied at all ? Or are all requests being rejected ?
https://github.com/istio/istio/wiki/Collecting-Logs-and-Debug-information#turn-on-debug-on-mixer
Turn on debug for API on mixer and send us logs from istio-policy.

**Comment by DaspawnW** [comment]
It seems that the requests are all in one quota and not in seperate per principal.

Attached you can find two requests with two different users. The first user has access and the second not because he is in the same quota cache.

First User principal: https://<removed for security reasons>/auth/realms/default/e8f6311e-57b0-49f2-bd84-b4e056ffd2f4
Second User principal: https://<removed for security reasons>/auth/realms/default/e92d9f8a-ae6d-407f-889f-15a1121f33fc

```
2018-09-07T21:01:07.488465Z	debug	api	Dispatching Check
2018-09-07T21:01:07.488475Z	debug	No destinations found for variety: table='66', variety='0'
2018-09-07T21:01:07.488480Z	debug	api	Check approved
2018-09-07T21:01:07.492400Z	debug	api	Check (GlobalWordCount:203, DeduplicationID:03560275-a5cd-4c78-9d84-0a5164eba5687525633059764136553, Quota:map[requestcount:{1 false}])
2018-09-07T21:01:07.492417Z	debug	api	Dispatching Preprocess Check
2018-09-07T21:01:07.492522Z	debug	attributes	Creating bag with attributes: connection.mtls               : true
context.protocol              : http
context.reporter.kind         : inbound
context.reporter.uid          : kubernetes://ms-1-6fdfdd5557-r5j44.microservices
destination.ip                : [0 0 0 0 0 0 0 0 0 0 255 255 10 200 0 53]
destination.namespace         : microservices
destination.port              : 8080
destination.principal         : cluster.local/ns/microservices/sa/default
destination.service           : ms-1.microservices.svc.cluster.local
destination.service.host      : ms-1.microservices.svc.cluster.local
destination.service.name      : ms-1
destination.service.namespace : microservices
destination.service.uid       : istio://microservices/services/ms-1
destination.uid               : kubernetes://ms-1-6fdfdd5557-r5j44.microservices
origin.ip                     : [10 200 0 26]
request.auth.audiences        : test-api
request.auth.claims           : {request.auth.claims map[<removed for security reasons>] 0xc4368b2420}
request.auth.presenter        : test-api
request.auth.principal        : https://<removed for security reasons>/auth/realms/default/e8f6311e-57b0-49f2-bd84-b4e056ffd2f4
request.auth.raw_claims       : {<removed for security reasons>}
request.headers               : {request.headers map[x-b3-sampled:1 x-original-uri:/api/ms1 x-b3-traceid:34019ecc565da288 sec-istio-authn-payload:<removed for security reasons> x-request-id:6927fa83-c7ba-966b-9bf9-d15b2d64102b x-b3-flags:0 user-agent:curl/7.35.0 x-forwarded-proto:https x-b3-parentspanid:1ed56b64def8ab39 x-real-ip:10.200.0.38 x-forwarded-port:443 content-length:0 x-forwarded-host:ms-1.example.play x-b3-spanid:cf4676fdf26e961f x-forwarded-for:10.200.0.38, 10.200.0.49 :method:GET accept:*/* :authority:ms-1.example.play x-scheme:https x-envoy-external-address:10.200.0.49 :path:/api/ms1 authorization:Bearer <removed for security reasons>] 0xc4368b2420}
request.host                  : ms-1.example.play
request.method                : GET
request.path                  : /api/ms1
request.scheme                : http
request.time                  : 2018-09-07 21:01:09.303934952 +0000 UTC
request.useragent             : curl/7.35.0
source.principal              : cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account
source.uid                    : kubernetes://istio-ingressgateway-5ccb8776dc-c5vr6.istio-system
source.user                   : cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account

2018-09-07T21:01:07.492554Z	debug	no rules for namespace, using defaults: table='66', variety='3', ns='microservices'
2018-09-07T21:01:07.492623Z	debug	begin dispatch: destination='kubernetes:handler.kubernetesenv.istio-system(kubernetesenv)'
2018-09-07T21:01:07.492690Z	debug	complete dispatch: destination='kubernetes:handler.kubernetesenv.istio-system(kubernetesenv)' {err:<nil>}
2018-09-07T21:01:07.492732Z	debug	api	Dispatching to main adapters after running processors
2018-09-07T21:01:07.492871Z	debug	api	Attribute Bag:
connection.mtls               : true
context.protocol              : http
context.reporter.kind         : inbound
context.reporter.uid          : kubernetes://ms-1-6fdfdd5557-r5j44.microservices
destination.ip                : [0 0 0 0 0 0 0 0 0 0 255 255 10 200 0 53]
destination.namespace         : microservices
destination.port              : 8080
destination.principal         : cluster.local/ns/microservices/sa/default
destination.service           : ms-1.microservices.svc.cluster.local
destination.service.host      : ms-1.microservices.svc.cluster.local
destination.service.name      : ms-1
destination.service.namespace : microservices
destination.service.uid       : istio://microservices/services/ms-1
destination.uid               : kubernetes://ms-1-6fdfdd5557-r5j44.microservices
origin.ip                     : [10 200 0 26]
request.auth.audiences        : test-api
request.auth.claims           : {<removed for security reasons>}
request.auth.presenter        : test-api
request.auth.principal        : https://<removed for security reasons>/auth/realms/default/e8f6311e-57b0-49f2-bd84-b4e056ffd2f4
request.auth.raw_claims       : {<removed for security reasons>}
request.headers               : {request.headers map[x-envoy-external-address:10.200.0.49 :path:/api/ms1 authorization:Bearer <removed for security reasons> accept:*/* :authority:ms-1.example.play x-scheme:https x-request-id:6927fa83-c7ba-966b-9bf9-d15b2d64102b x-b3-flags:0 user-agent:curl/7.35.0 x-b3-sampled:1 x-original-uri:/api/ms1 x-b3-traceid:34019ecc565da288 sec-istio-authn-payload:<removed for security reasons> x-forwarded-port:443 content-length:0 x-forwarded-proto:https x-b3-parentspanid:1ed56b64def8ab39 x-real-ip:10.200.0.38 :method:GET x-forwarded-host:ms-1.example.play x-b3-spanid:cf4676fdf26e961f x-forwarded-for:10.200.0.38, 10.200.0.49] 0xc4368b2420}
request.host                  : ms-1.example.play
request.method                : GET
request.path                  : /api/ms1
request.scheme                : http
request.time                  : 2018-09-07 21:01:09.303934952 +0000 UTC
request.useragent             : curl/7.35.0
source.principal              : cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account
source.uid                    : kubernetes://istio-ingressgateway-5ccb8776dc-c5vr6.istio-system
source.user                   : cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account
---
destination.container.name    : ms-1
destination.labels            : map[app:ms-1 pod-template-hash:2989881113]
destination.name              : ms-1-6fdfdd5557-r5j44
destination.owner             : kubernetes://apis/apps/v1/namespaces/microservices/deployments/ms-1
destination.serviceAccount    : default
destination.workload.name     : ms-1
destination.workload.namespace: microservices
destination.workload.uid      : istio://microservices/workloads/ms-1
source.ip                     : [10 200 0 26]
source.labels                 : map[app:istio-ingressgateway istio:ingressgateway pod-template-hash:1776433287]
source.name                   : istio-ingressgateway-5ccb8776dc-c5vr6
source.namespace              : istio-system
source.owner                  : kubernetes://apis/apps/v1/namespaces/istio-system/deployments/istio-ingressgateway
source.serviceAccount         : istio-ingressgateway-service-account
source.workload.name          : istio-ingressgateway
source.workload.namespace     : istio-system
source.workload.uid           : istio://istio-system/workloads/istio-ingressgateway

2018-09-07T21:01:07.492887Z	debug	api	Dispatching Check
2018-09-07T21:01:07.492896Z	debug	No destinations found for variety: table='66', variety='0'
2018-09-07T21:01:07.492901Z	debug	api	Check approved
2018-09-07T21:01:07.492924Z	debug	api	Dispatching Quota: requestcount
2018-09-07T21:01:07.492931Z	debug	no rules for namespace, using defaults: table='66', variety='2', ns='microservices'
2018-09-07T21:01:07.492966Z	debug	begin dispatch: destination='quota:handler.redisquota.istio-system(redisquota)'
2018-09-07T21:01:07.492980Z	debug	adapters	key: requestcount.quota.istio-system maxAmount: 1	{"adapter": "handler.redisquota.istio-system"}
2018-09-07T21:01:07.494481Z	debug	complete dispatch: destination='quota:handler.redisquota.istio-system(redisquota)' {err:<nil>}
2018-09-07T21:01:07.494522Z	debug	api	Quota 'requestcount' result: v1.CheckResponse_QuotaResult{ValidDuration:20000000000, GrantedAmount:1, ReferencedAttributes:v1.ReferencedAttributes{Words:[]string(nil), AttributeMatches:[]v1.ReferencedAttributes_AttributeMatch(nil)}}
2018-09-07T21:01:09.778265Z	debug	istio.io/istio/mixer/adapter/kubernetesenv/cache.go:145: Watch close - *v1.Pod total 0 items received
2018-09-07T21:01:11.208946Z	debug	api	Check (GlobalWordCount:203, DeduplicationID:03560275-a5cd-4c78-9d84-0a5164eba5687525633059764136554, Quota:map[requestcount:{10 true}])
2018-09-07T21:01:11.208976Z	debug	api	Dispatching Preprocess Check
2018-09-07T21:01:11.209102Z	debug	attributes	Creating bag with attributes: check.cache_hit               : true
connection.mtls               : true
context.protocol              : http
context.reporter.kind         : inbound
context.reporter.uid          : kubernetes://ms-1-6fdfdd5557-r5j44.microservices
destination.ip                : [0 0 0 0 0 0 0 0 0 0 255 255 10 200 0 53]
destination.namespace         : microservices
destination.port              : 8080
destination.principal         : cluster.local/ns/microservices/sa/default
destination.service           : ms-1.microservices.svc.cluster.local
destination.service.host      : ms-1.microservices.svc.cluster.local
destination.service.name      : ms-1
destination.service.namespace : microservices
destination.service.uid       : istio://microservices/services/ms-1
destination.uid               : kubernetes://ms-1-6fdfdd5557-r5j44.microservices
origin.ip                     : [10 200 0 26]
quota.cache_hit               : true
request.auth.audiences        : test-api
request.auth.claims           : {<removed for security reasons>}
request.auth.presenter        : test-api
request.auth.principal        : https://<removed for security reasons>/auth/realms/default/e92d9f8a-ae6d-407f-889f-15a1121f33fc
request.auth.raw_claims       : {<removed for security reasons>}
request.headers               : {request.headers map[x-b3-parentspanid:c5dcad37ae33a331 x-b3-flags:0 x-b3-sampled:1 :authority:ms-1.example.play x-forwarded-proto:https :path:/api/ms1 x-scheme:https content-length:0 accept:*/* authorization:Bearer <removed for security reasons> user-agent:curl/7.35.0 x-forwarded-host:ms-1.example.play :method:GET x-original-uri:/api/ms1 x-b3-traceid:7d5bf7cea838e2c7 x-envoy-external-address:10.200.0.49 x-real-ip:10.200.0.38 x-b3-spanid:771883934f0f4cf9 x-request-id:37b6d066-3145-9035-b89c-ecc8a1fed955 x-forwarded-for:10.200.0.38, 10.200.0.49 x-forwarded-port:443 sec-istio-authn-payload:<removed for security reasons>] 0xc428332060}
request.host                  : ms-1.example.play
request.method                : GET
request.path                  : /api/ms1
request.scheme                : http
request.time                  : 2018-09-07 21:01:13.021201421 +0000 UTC
request.useragent             : curl/7.35.0
source.principal              : cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account
source.uid                    : kubernetes://istio-ingressgateway-5ccb8776dc-c5vr6.istio-system
source.user                   : cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account

2018-09-07T21:01:11.209137Z	debug	Check approved from cache
2018-09-07T21:01:11.209149Z	debug	no rules for namespace, using defaults: table='66', variety='3', ns='microservices'
2018-09-07T21:01:11.209222Z	debug	begin dispatch: destination='kubernetes:handler.kubernetesenv.istio-system(kubernetesenv)'
2018-09-07T21:01:11.209312Z	debug	complete dispatch: destination='kubernetes:handler.kubernetesenv.istio-system(kubernetesenv)' {err:<nil>}
2018-09-07T21:01:11.209374Z	debug	api	Dispatching to main adapters after running processors
2018-09-07T21:01:11.209481Z	debug	api	Attribute Bag:
check.cache_hit               : true
connection.mtls               : true
context.protocol              : http
context.reporter.kind         : inbound
context.reporter.uid          : kubernetes://ms-1-6fdfdd5557-r5j44.microservices
destination.ip                : [0 0 0 0 0 0 0 0 0 0 255 255 10 200 0 53]
destination.namespace         : microservices
destination.port              : 8080
destination.principal         : cluster.local/ns/microservices/sa/default
destination.service           : ms-1.microservices.svc.cluster.local
destination.service.host      : ms-1.microservices.svc.cluster.local
destination.service.name      : ms-1
destination.service.namespace : microservices
destination.service.uid       : istio://microservices/services/ms-1
destination.uid               : kubernetes://ms-1-6fdfdd5557-r5j44.microservices
origin.ip                     : [10 200 0 26]
quota.cache_hit               : true
request.auth.audiences        : test-api
request.auth.claims           : {<removed for security reasons>}
request.auth.presenter        : test-api
request.auth.principal        : https://<removed for security reasons>/auth/realms/default/e92d9f8a-ae6d-407f-889f-15a1121f33fc
request.auth.raw_claims       : {<removed for security reasons>}
request.headers               : {request.headers map[x-original-uri:/api/ms1 x-b3-traceid:7d5bf7cea838e2c7 x-envoy-external-address:10.200.0.49 x-real-ip:10.200.0.38 x-b3-spanid:771883934f0f4cf9 :method:GET x-forwarded-for:10.200.0.38, 10.200.0.49 x-forwarded-port:443 sec-istio-authn-payload:<removed for security reasons> x-request-id:37b6d066-3145-9035-b89c-ecc8a1fed955 x-b3-flags:0 x-b3-sampled:1 :authority:ms-1.example.play x-b3-parentspanid:c5dcad37ae33a331 x-scheme:https content-length:0 accept:*/* authorization:Bearer <removed for security reasons> user-agent:curl/7.35.0 x-forwarded-host:ms-1.example.play x-forwarded-proto:https :path:/api/ms1] 0xc428332060}
request.host                  : ms-1.example.play
request.method                : GET
request.path                  : /api/ms1
request.scheme                : http
request.time                  : 2018-09-07 21:01:13.021201421 +0000 UTC
request.useragent             : curl/7.35.0
source.principal              : cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account
source.uid                    : kubernetes://istio-ingressgateway-5ccb8776dc-c5vr6.istio-system
source.user                   : cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account
---
destination.container.name    : ms-1
destination.labels            : map[app:ms-1 pod-template-hash:2989881113]
destination.name              : ms-1-6fdfdd5557-r5j44
destination.owner             : kubernetes://apis/apps/v1/namespaces/microservices/deployments/ms-1
destination.serviceAccount    : default
destination.workload.name     : ms-1
destination.workload.namespace: microservices
destination.workload.uid      : istio://microservices/workloads/ms-1
source.ip                     : [10 200 0 26]
source.labels                 : map[istio:ingressgateway pod-template-hash:1776433287 app:istio-ingressgateway]
source.name                   : istio-ingressgateway-5ccb8776dc-c5vr6
source.namespace              : istio-system
source.owner                  : kubernetes://apis/apps/v1/namespaces/istio-system/deployments/istio-ingressgateway
source.serviceAccount         : istio-ingressgateway-service-account
source.workload.name          : istio-ingressgateway
source.workload.namespace     : istio-system
source.workload.uid           : istio://istio-system/workloads/istio-ingressgateway

2018-09-07T21:01:11.209493Z	debug	api	Dispatching Check
2018-09-07T21:01:11.209504Z	debug	No destinations found for variety: table='66', variety='0'
2018-09-07T21:01:11.209508Z	debug	api	Check approved
2018-09-07T21:01:11.209530Z	debug	api	Dispatching Quota: requestcount
2018-09-07T21:01:11.209536Z	debug	no rules for namespace, using defaults: table='66', variety='2', ns='microservices'
2018-09-07T21:01:11.209557Z	debug	begin dispatch: destination='quota:handler.redisquota.istio-system(redisquota)'
2018-09-07T21:01:11.209572Z	debug	adapters	key: requestcount.quota.istio-system maxAmount: 1	{"adapter": "handler.redisquota.istio-system"}
2018-09-07T21:01:11.211306Z	debug	complete dispatch: destination='quota:handler.redisquota.istio-system(redisquota)' {err:<nil>}
2018-09-07T21:01:11.211335Z	debug	api	Quota denied: {8 handler.redisquota.istio-system:redisquota: Resource exhausted []}
2018-09-07T21:01:11.211349Z	debug	api	Quota 'requestcount' result: v1.CheckResponse_QuotaResult{ValidDuration:0, GrantedAmount:0, ReferencedAttributes:v1.ReferencedAttributes{Words:[]string(nil), AttributeMatches:[]v1.ReferencedAttributes_AttributeMatch(nil)}}
```

**Comment by mandarjog** [comment]
@bianpengyuan can you please take a look at this?
@DaspawnW are you able to replicate this by using `memquota` adapter?

**Comment by denniseffing** [comment]
I noticed the same behaviour when using `request.headers["x-forwarded-for"]`.
Requests **are not** rate limited per client IP at all, even though the documentation clearly says so. I assume that there is some kind of general issue regarding the request counters. 

The memquota adapter also seems to introduce some inconsistencies. Some requests are denied even though they should be allowed.
See https://groups.google.com/forum/#!topic/istio-users/iWOPtcUHINs for further information.

**Comment by bianpengyuan** [comment]
There are several separate issues:

@DaspawnW in the initial issue that you posted, the `source` dimension is configured as `source: request.headers["X-Test-Header"] | "unknown"`, however header map key are all lower cased, which essentially makes source `unknown` in all instances and explains behavior you observed (quota enforcement cannot distinguish different test header). For the second issue with authorization header or auth principle, I am not sure what do you mean by "won't work". What do you observe exactly? Based on log seems like the quota adapter is invoked correctly.

@denniseffing I think this works as expected.. @mandarjog correct me if I'm wrong. Envoy starts multiple threads based on number of cores. Each thread only deals with one request at a time and also each thread instantiates a mixer filter (mixer client) to talk to mixer. Mixer client runs a prefetch algorithm which by default will fetch 10 tokens at the beginning, and exhausts the configured quota amount (5 per 5s). All other threads except the first thread cannot fetch any quota, which causes 429 even for the first five requests. I experimented with proxy concurrency as 1 and this behavior is gone.

**Comment by denniseffing** [comment]
@bianpengyuan Rate limiting per client IP (based on `request.headers["x-forwarded-for"]` as outlined in the Istio docs) doesn't work at all. 

I tested this with the following shell script executing simultaneously on two different machines:
```shell
while true; do
  curl -o /dev/null --silent --head --write-out "$(date +"%H:%M:%S.%3N") %{http_code}\n" $1
done
```

Only 5 requests per 5 seconds are allowed **overall**, not for each machine.

I turned on debugging in Mixer to take a look at the attribute values:

**request.headers map**
```
{
  request.headers map[
    upgrade-insecure-requests:1 
    x-forwarded-proto:http 
    accept-encoding:gzip, deflate 
    accept:text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8 
    x-envoy-internal:true 
    cache-control:max-age=0 
    :method:GET 
    x-forwarded-for:172.17.0.1 
    user-agent:Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36 
    x-b3-traceid:64122dcb2ab6fce0 
    x-b3-spanid:64122dcb2ab6fce0 
    accept-language:de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7 
    content-length:0 :path:/productpage 
    x-request-id:bec0cbf7-f54f-946b-8655-a8968d0dd4db 
    :authority:192.168.2.91:31380 x-b3-sampled:1
  ] 0xc4294c00c0
}
```

**Selected quota override**
```
quota override: 
{
  map[
    destination:productpage
  ] 5 5s
} selected for {
  requestcount.quota.istio-system 
    map[
      destination:productpage 
      destinationVersion:v1 
      source:172.17.0.1
  ]
}	
{"adapter": "handler.memquota.istio-system"}
```

I don't know where the IP `172.17.0.1` for `x-forwarded-for` comes from, but it is definitely not the IP of the machine that made the request to the productpage in the first place. It seems to be a Cluster IP but I couldn't find a pod that uses it.

The IP is the same, regardless of the origin of the initial request. That also explains why the rate limit is **not** applied for each client IP.

**Comment by bianpengyuan** [comment]
@denniseffing That is your kube-proxy's IP

**Comment by denniseffing** [comment]
@bianpengyuan Yup, that makes sense. Do you have any thoughts on the rate limit per client IP issue?

**Comment by DaspawnW** [comment]
Hi @bianpengyuan,

here a short example:
User 1 sends requests with a principal x (`request.auth.principal`). If his quota is exceeded he can't send any requests (as expected ;-) )
User 2 sends requests with a principal y (`request.auth.principal`). He is directly denied without any successful request.

**Comment by denniseffing** [comment]
The bug in rate limiting per client IP is directly related to #7679, #7328 and #7918.

**Comment by DaspawnW** [comment]
Hi @denniseffing,

It seems to be related to the x-forwarded-ip but not to the custom header, or?

**Comment by denniseffing** [comment]
@DaspawnW When I first commented in this issue, I thought that your `request.auth.principal` issue is directly related to the rate limiting per client IP issue, **but it is not**. I didn't know that the `x-forwarded-for` header contains the wrong IP address (which is stated in #7679 and explains why rate limiting per client IP doesn't work).


I hope this clarifies the misunderstanding.

**Comment by vbuciuc** [comment]
Looking at the code, there is no counter per "Cartesian product of all quota dimensions" (what does that even mean?) but rather one counter per quota template + a separate counter for each dimension override in the handler definition (not really, since only keys are used to build the hash for the counter name so I see a possibility that two overrides of the same dimension might conflict as well).

https://github.com/istio/istio/blob/master/mixer/adapter/redisquota/redisquota.go#L289

In short there is no way you can do quota per specific user unless you know all the users when defining the handler and even then it might act weird.
